lupa/_lupa.pyx

```diff
@@ -2157,6 +2220,197 @@ cdef int py_set_overflow_handler(lua_State* L) nogil:
     lua.lua_setfield(L, lua.LUA_REGISTRYINDEX, LUPAOFH)  #
     return 0
 
+# type checking for Python objects in Lua
+
+cdef int py_is_error_with_gil(lua_State* L, py_object* py_obj) with gil:
+    cdef LuaRuntime runtime = None
+    try:
+        runtime = <LuaRuntime?>py_obj.runtime
+        obj = <object>py_obj.obj
+        lua.lua_pushboolean(L, isinstance(obj, _PyException))
+        return 1
+    except:
+        return py_to_lua_error(runtime, L, b'error checking if object is Python error')
+
+cdef int py_is_error(lua_State* L) nogil:
+    cdef py_object* py_obj = unpack_userdata(L, 1)
+    if not py_obj:
+        lua.lua_pushboolean(L, 0)
+        return 1
+    result = py_is_error_with_gil(L, py_obj)
+    if result < 0:
+        return lua.lua_error(L) # never returns!
+    return result
+
+cdef int py_is_object(lua_State* L) nogil:
+    cdef py_object* py_obj
+    if lua.lua_isuserdata(L, 1):
+        py_obj = unpack_userdata(L, 1)
+    else:
+        py_obj = unpack_wrapped_pyfunction(L, 1)
+    lua.lua_pushboolean(L, py_obj != NULL)
+    return 1
+
+# raising Python errors from Lua
+
+cdef object tb_set_next(object tb, object tb_next):
+    c_tb = <PyTracebackObject*>tb
+    if tb.tb_next is not None:
+        prev_tb_next = <object>c_tb.tb_next
+        c_tb.tb_next = NULL
+        cpython.ref.Py_DECREF(prev_tb_next)
+    if tb_next is not None:
+        cpython.ref.Py_INCREF(tb_next)
+        c_tb.tb_next = <PyTracebackObject*>tb_next
+    return tb
+
+cdef object fake_traceback(object exc, object filename, object name, int lineno):
+    scope = {
+            "__name__": filename,
+            "__file__": filename,
+            "__lupa_exception__": exc,
+    }
+    code = compile("\n" * (lineno - 1) + "raise __lupa_exception__", filename, "exec")
```

Comment:
You're right, the `__Pyx_AddTraceback` does something with Frame and Code objects. I will look into that.

Reply:
So, I took a look at the code generated by Cython, and it's very messy. I think this solution in pure Python is more portable (although not optimal). Could we leave this improvement for a future PR? It's not a big performance bottleneck because, unless someone is constantly throwing and catching errors in a tight loop, code objects and frames will be built on rare occasions.

Reply:
There was a related discussion going on on `python-dev`, so I asked and there seems to be general interest in improving the C-API for this. Not something for right now, but once that's available and used (and backported) by Cython, Lupa (and probably several other projects that handle foreign languages, DSLs, templates, …) could benefit from it.

https://mail.python.org/archives/list/python-dev@python.org/thread/ZWTBR5ESYR26BUIVMXOKPFRLGGYDJSFC/#XYNNMH57O7CYWHYKTD3ELZTM3B4M53HL